airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py

```diff
@@ -389,10 +389,23 @@ def read_records(
         # Warning: use self.state instead of the stream_state passed as argument!
         stream_slice = stream_slice or {}  # None-check
         self.paginator.reset()
+        # Note: Adding the state per partition led to a difficult situation where the state for a partition is not the same as the
+        # stream_state. This means that if any class downstream wants to access the state, it would need to perform some kind of selection
+        # based on the partition. To short circuit this, we do the selection here which avoid downstream classes to know about it the
+        # partition. We have generified the problem to the stream slice instead of the partition because it is the level of abstraction
+        # streams know (they don't know about partitions). However, we're still unsure as how it will evolve since we can't see any other
+        # cursor doing selection per slice. We don't want to pollute the interface. Therefore, we will keep the `hasattr` hack for now.
+        # * What is the information we need to clean the hasattr? Once we will have another case where we need to select a state, we will
+        #    know if the abstraction using `stream_slice` so select to state is the right one and validate if the interface makes sense.
+        # * Why is this abstraction not on the DeclarativeStream level? DeclarativeStream does not have a notion of stream slicers. However,
+        #    it might make more sense to have this on the stream level since this logic would probably apply to the other retrievers. As of
+        #    2023-06-09, this is a breaking change because it would mean adding `stream_slicer` to the Retriever interface which we would
+        #    to avoid for now.
+        slice_state = self.stream_slicer.select(stream_slice, stream_state) if hasattr(self.stream_slicer, "select") else stream_state
         records_generator = self._read_pages(
             self.parse_records,
             stream_slice,
-            stream_state,
+            slice_state,
```

Comment:
Does that means that I need to rename all downstream classes from `stream_state` to `slice_state`? I feel like I should but it seems non trivial so I would like to validate before doing so

Reply:
Actually, I think this is the same concern as https://github.com/airbytehq/airbyte/pull/27223/files#diff-574e7f982441af95b199fa0d628677b9a654543eaea1534b7516c8679375a235R392 and maybe we should wait...